### PR TITLE
Manage Package page forms should link back to the Manage page

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/nuget/nugetgallery/issues/7567
https://github.com/nuget/nugetgallery/issues/7495

Currently, when you submit a form on the Manage Package page (Deprecation, Listing, Documentation) you are linked back to the package details page. This is annoying when you are trying to do other management operations at the same time.

This changes it so that submitting the forms stays on the Manage Package page. Additionally, I added "TempData" messages so that it's clearer what operation was performed.

### CODE REVIEW

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 1]
```diff
@@ -3627,15 +3627,14 @@ public async Task UpdatesUnlistedIfSelected(User currentUser, User owner)
...
 
                 // Act
-                var result = await controller.Edit("Foo", "1.0", listed: false, urlFactory: (pkg, relativeUrl) => @"~\Bar.cshtml");
+                var result = await controller.UpdateListed("Foo", "1.0", false);
 
                 // Assert
                 packageService.Verify();
                 Assert.IsType<RedirectResult>(result);
-                Assert.Equal(@"~\Bar.cshtml", ((RedirectResult)result).Url);
             }
```
Comment (zhhyu) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Is that possible to verify the redirect Url here?

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 2]
```diff
@@ -3647,6 +3646,7 @@ public async Task UpdatesUnlistedIfNotSelected(User currentUser, User owner)
                 {
                     PackageRegistration = new PackageRegistration { Id = "Foo" },
                     Version = "1.0",
+                    NormalizedVersion = "1.0.0",
```
Comment (zhhyu) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: What is the purpose to introduce this normalizedVersion here?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 3]
```diff
@@ -1969,9 +2000,11 @@ public virtual async Task<JsonResult> Edit(string id, string version, VerifyPack
                 }
             }
 
+            TempData["Message"] = "Your package's documentation has been updated.";
+
             return Json(new
             {
-                location = returnUrl ?? Url.Package(id, version)
+                location = returnUrl ?? Url.ManagePackage(new TrivialPackageVersionModel(id, version))
```
Comment (zhhyu) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Regarding this change on the Edit method, do we have any test update to reflect this change?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 4]
```diff
@@ -1905,8 +1905,39 @@ public virtual ActionResult Edit(string id, string version)
         [ValidateAntiForgeryToken]
         public virtual async Task<ActionResult> UpdateListed(string id, string version, bool? listed)
         {
-            // Edit does exactly the same thing that Delete used to do... REUSE ALL THE CODE!
-            return await Edit(id, version, listed, Url.Package);
+            var package = _packageService.FindPackageByIdAndVersionStrict(id, version);
+            if (package == null)
+            {
```
Comment (zhhyu) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Maybe it's better that we have the test coverage on this case.

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 5]
```diff
@@ -1905,8 +1905,39 @@ public virtual ActionResult Edit(string id, string version)
...
+            }
+            else
+            {
+                action = "listed";
+                await _packageUpdateService.MarkPackageListedAsync(package);
+            }
+            TempData["Message"] = string.Format(
+                CultureInfo.CurrentCulture,
+                "The package has been {0}. It may take several hours for this change to propagate through our system.",
+                action);
```
Comment (zhhyu) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Maybe it's better that we also verify this TempData in the tests.

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs (hunk 2 above)
Comment (scottbommarito) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: `TrivialPackageViewModel(Package)` uses `Package.NormalizedVersion`, so we need to specify it or else the URL generation in the test throws a null-ref.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 4 above)
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/PackagesController.cs]: This is already covered by tests. `UpdateListed` used to be a one-liner that called a function that was used nowhere else, so I moved all the code in the function up into `UpdateListed`.

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs (hunk 1 above)
Comment (scottbommarito) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: We typically don't for `UrlHelper.*` redirects, because the way the URLs are generated in the tests is sometimes unexpected or broken.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/PackagesController.cs]: No, as mentioned above, because `UrlHelper.*` calls are difficult to verify.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 4 above)
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/PackagesController.cs]: Actually, good call, we don't have any tests for missing package here. I can add one.

File: src/NuGetGallery/Controllers/ManageDeprecationJsonApiController.cs [hunk 6]
```diff
@@ -53,6 +53,13 @@ public virtual JsonResult GetAlternatePackageVersions(string id)
                 return Json(error.Status, new { error = error.Message });
             }
 
+            var packagePluralString = request.Versions.Count() > 1 ? "packages have" : "package has";
+            var deprecatedString = request.IsLegacy || request.HasCriticalBugs || request.IsOther
+                ? "deprecated" : "undeprecated";
+            TempData["Message"] =
```
Comment (zhhyu) [inline src/NuGetGallery/Controllers/ManageDeprecationJsonApiController.cs]: Maybe here also verify the TempData in the test? :smile:

Comment (scottbommarito) [reply src/NuGetGallery/Controllers/ManageDeprecationJsonApiController.cs]: Already did haha: https://github.com/NuGet/NuGetGallery/pull/7678/files#diff-d8fa5513996103b3cf41923124b0defbR109
